lightning/glue/glue.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package glue
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb-lightning/lightning/checkpoints"
+	"github.com/pingcap/tidb-lightning/lightning/common"
+	"github.com/pingcap/tidb-lightning/lightning/config"
+	"github.com/pingcap/tidb-lightning/lightning/log"
+	"go.uber.org/zap"
+)
+
+type Glue interface {
+	OwnsSQLExecutor() bool
+	GetSQLExecutor() SQLExecutor
+	GetParser() *parser.Parser
+	GetTables(context.Context, string) ([]*model.TableInfo, error)
+	OpenCheckpointsDB(context.Context, *config.Config) (checkpoints.CheckpointsDB, error)
+	Record(string, uint64)
+}
+
+type SQLExecutor interface {
+	ExecuteWithLog(ctx context.Context, query string, purpose string, logger *zap.Logger) error
+	ObtainStringLog(ctx context.Context, query string, purpose string, logger *zap.Logger) (string, error)
+	Close()
+}
```

Comment:
Why pass a logger through the interface?

What about passing a logger when we creating an SQLExecutor instance?

```golang
func NewExternalTiDBGlue(db *sql.DB, sqlMode mysql.SQLMode, logger *zap.Logger) *ExternalTiDBGlue { ... }
```

Reply:
some functions need a logger from `Task`, and that `Task` is often built from logger with different fields. It seems that pinning a logger at `New` is not very flexible